lib/stdio/direction.js

```diff
@@ -0,0 +1,42 @@
+import {
+	isStream as isNodeStream,
+	isReadableStream as isNodeReadableStream,
+	isWritableStream as isNodeWritableStream,
+} from 'is-stream';
+import {isWritableStream} from './type.js';
+
+// For `stdio[index]` beyond stdin/stdout/stderr, we need to guess whether the value passed is intended for inputs or outputs.
+// This allows us to know whether to pipe _into_ or _from_ the stream.
+export const addStreamDirection = stdioStream => {
+	const direction = getStreamDirection(stdioStream);
+	return addDirection(stdioStream, direction);
+};
+
+const getStreamDirection = stdioStream => KNOWN_DIRECTIONS[stdioStream.index] ?? guessStreamDirection[stdioStream.type](stdioStream.value);
+
+// `stdin`/`stdout`/`stderr` have a known direction
+const KNOWN_DIRECTIONS = ['input', 'output', 'output'];
+
+// `stringOrBuffer` type always applies to `stdin`, i.e. does not need to be handled here
+const guessStreamDirection = {
+	filePath: () => undefined,
+	iterable: () => 'input',
+	webStream: stdioOption => isWritableStream(stdioOption) ? 'output' : 'input',
+	nodeStream(stdioOption) {
+		if (isNodeReadableStream(stdioOption)) {
+			return isNodeWritableStream(stdioOption) ? undefined : 'input';
+		}
+
+		return 'output';
+	},
+	native(stdioOption) {
+		if (isNodeStream(stdioOption)) {
+			return guessStreamDirection.nodeStream(stdioOption);
+		}
+	},
+};
```

Comment:
The "direction" is whether the `stdio` option value is meant for input or output. For example, a readable vs a writable stream.

Some directions are ambiguous, e.g. a file path can be used both for reads or writes. In that case, we use `undefined`, which eventually defaults to `output`.